crates/nu-command/src/filesystem/cp.rs

```diff
@@ -345,29 +377,97 @@ fn interactive_copy(
     }
 }
 
+// `copy_file` uses `std::fs::copy` to copy a file. There is another function called `copy_file_with_progressbar`
+// which uses `read` and `write` instead. This is to get the progress of the copy. If you make any changes in this
+// function try to match the changes in `copy_file_with_progressbar`
 fn copy_file(src: PathBuf, dst: PathBuf, span: Span) -> Value {
     match std::fs::copy(&src, &dst) {
         Ok(_) => {
             let msg = format!("copied {:} to {:}", src.display(), dst.display());
             Value::String { val: msg, span }
         }
-        Err(e) => {
-            let message = format!("copy file {src:?} failed: {e}");
-
-            use std::io::ErrorKind;
-            let shell_error = match e.kind() {
-                ErrorKind::NotFound => ShellError::FileNotFoundCustom(message, span),
-                ErrorKind::PermissionDenied => ShellError::PermissionDeniedError(message, span),
-                ErrorKind::Interrupted => ShellError::IOInterrupted(message, span),
-                ErrorKind::OutOfMemory => ShellError::OutOfMemoryError(message, span),
-                // TODO: handle ExecutableFileBusy etc. when io_error_more is stabilized
-                // https://github.com/rust-lang/rust/issues/86442
-                _ => ShellError::IOErrorSpanned(message, span),
-            };
-
-            Value::Error { error: shell_error }
-        }
+        Err(e) => error_return(e, src, span),
+    }
+}
+
+// `copy_file_with_progressbar` uses `read` and `write` to copy a file. There is another function called `copy_file`
+// which uses `std::fs::copy` instead which is faster but can't gt the process of the copy. If you make any changes
+// in this function try to match the changes in `copy_file`
+fn copy_file_with_progressbar(src: PathBuf, dst: PathBuf, span: Span) -> Value {
+    let mut bytes_processed: u64 = 0;
+    let bytes_processed_p = &mut bytes_processed;
+    let mut process_failed: Option<std::io::Error> = None;
+    let process_failed_p = &mut process_failed;
+
+    let mut file_in = match std::fs::File::open(&src) {
+        Ok(file) => file,
+        Err(error) => return error_return(error, src, span),
+    };
+
+    let file_size = match file_in.metadata() {
+        Ok(metadata) => Some(metadata.len()),
+        _ => None,
+    };
+
+    let bar = progress_bar::NuProgressBar::new(file_size);
+    let mut bar_clone = bar.clone();
+
+    let mut file_out = match std::fs::File::create(&dst) {
+        Ok(file) => file,
+        Err(error) => return error_return(error, src, span),
+    };
+    let mut buffer = [0u8; 8192];
```

Comment:
Have you considered using `BufReader`/`BufWriter`? They will handle buffering for you

Reply:
Hi @rgwood! I didn't know about [fs_extra](https://github.com/webdesus/fs_extra). 
I just looked at their source code and there is not much difference in how they implemented it. There is just one thing I noticed that I wanted to asked you about (and I am sorry if it very dumb or makes not sense)  they are defining the 
buffer as a vector `vec![0; options.buffer_size]` I in the other hand `[0u8; 8192]` could it be possible that vec is slower?
https://github.com/webdesus/fs_extra/blob/bb12c0d7cc286614fad8b59b8d9961fac8b7bde2/src/file.rs#L197

I also noticed you have an open issue there for 4 months. Is that still a problem? also they are not using `BufReader`/`BufWriter` so if we decide to not use `fs_extra` I can implement those.

Those are my thoughts. I am not against using it and if you think that is the best I will just make the changes. I also going to add some test if possible. 

Let me now what you think :)

Reply:
> could it be possible that vec is slower?

I don't think it would make a meaningful difference but I can't say for sure without measuring.

> I also noticed you have an open issue there for 4 months. Is that still a problem?

It's a huge problem. I'm really unimpressed that the maintainer still has not fixed [this critical bug](https://github.com/webdesus/fs_extra/issues/64). But I don't know if we'll get around to replacing `fs_extra` anytime soon.

Reply:
After further thought, I think I would vote for avoiding `fs_extra`.

Let's add some tests and use `BufReader` and `BufWriter`

Reply:
your vote is my command captain! I'll proceed!